gensim/similarities/termsim.py

```diff
@@ -293,17 +353,43 @@ def inner_product(self, X, Y, normalized=False):
         References
         ----------
         The soft cosine measure was perhaps first described by [sidorovetal14]_.
+        Further notes on the efficient implementation of the soft cosine measure are described by
+        [novotny18]_.
 
         .. [sidorovetal14] Grigori Sidorov et al., "Soft Similarity and Soft Cosine Measure: Similarity
            of Features in Vector Space Model", 2014, http://www.cys.cic.ipn.mx/ojs/index.php/CyS/article/view/2043/1921.
 
+        .. [novotny18] Vít Novotný, "Implementation Notes for the Soft Cosine Measure", 2018,
+           http://dx.doi.org/10.1145/3269206.3269317.
+
         """
         if not X or not Y:
             return self.matrix.dtype.type(0.0)
 
+        if normalized in (True, False):
+            warnings.warn(
+                'Boolean parameter normalized will be removed in 4.0.0, use '
+                'normalized=({normalized}, {normalized}) instead of '
+                'normalized={normalized}'.format(normalized=normalized),
+                category=DeprecationWarning,
+            )
+            normalized = (normalized, normalized)
+
+        normalized_X, normalized_Y = normalized
+        valid_normalized_values = (True, False, 'maintain')
+
+        if normalized_X not in valid_normalized_values:
+            raise ValueError('{} is not a valid value of normalize'.format(normalized_X))
+        if normalized_Y not in valid_normalized_values:
+            raise ValueError('{} is not a valid value of normalize'.format(normalized_Y))
+
         is_corpus_X, X = is_corpus(X)
         is_corpus_Y, Y = is_corpus(Y)
 
+        non_negative_norm_assertion_message = u"sparse documents must not contain any explicit " \
```

Comment:
This can be a internal (module-level) constant outside of the function.

Reply:
Fixed in dc26e56ad4b3f707c94d3b542aecb64996fc170b.